BPA/Ch.12/Nmix2.stan

```diff
@@ -1,145 +1,171 @@
 // Binomial-mixture model with overdispersion in both abundance and detection
+
+functions {
+  /**
+   * Returns log likelihood of N-mixture model
+   * with 2 replicated observations using
+   * bivariate Poisson distibution
+   *
+   * References
+   * Dennis et al. (2015) Computational aspects of N-mixture models.
+   *   Biometrics 71:237--246. DOI:10.1111/biom.12246
+   * Stan users mailing list
+   *   https://groups.google.com/forum/#!topic/stan-users/9mMsp1oB69g
+   *
+   * @param n          Number of observed individuals
+   * @param log_lambda Log of Poisson mean of population size
+   * @param logit_p    Logit of detection probability
+   *
+   * return Log probability
+  */
+  real bivariate_poisson_lpmf(int[] n, real log_lambda, real[] logit_p) {
+    real s[min(n) + 1];
+    real lambda = exp(log_lambda);
+    real p[2] = inv_logit(logit_p);
+
+    if (lambda < 0) {
+      reject("lambda must be non-negative.");
+    } else if (p[1] < 0 || p[1] > 1 || p[2] < 0 || p[2] > 1) {
+      reject("p must be in [0,1].");
+    }
+    for (u in 0:min(n))
+      s[u + 1] = poisson_lpmf(n[1] - u | lambda * p[1] * (1 - p[2]))
+        + poisson_lpmf(n[2] - u | lambda * p[2] * (1 - p[1]))
+        + poisson_lpmf(u | lambda * p[1] * p[2]);
+    return log_sum_exp(s);
+  }
+}
+
 data {
   int<lower=1> R;                // Number of sites
-  int<lower=1> T;                // Number of replications
-  int<lower=-1> y[R, T, 7];      // Counts (-1:NA)
+  // int<lower=1> T;                // Number of replications; fixed as 2
+  int<lower=-1> y[R, 2, 7];      // Counts (-1:NA)
   int<lower=1,upper=7> first[R]; // First occasion
   int<lower=1,upper=7> last[R];  // Last occasion
   int<lower=0> K;                // Upper bounds of population size
 }
 
 transformed data {
   int<lower=0> max_y[R, 7];
+  int<lower=0,upper=R> num_obs_site[7];
+  int T = 2;
 
   for (i in 1:R) {
     for (k in 1:(first[i] - 1))
-      max_y[i, k] <- 0;
+      max_y[i, k] = 0;
     for (k in (last[i] + 1):7)
-      max_y[i, k] <- 0;
+      max_y[i, k] = 0;
     for (k in first[i]:last[i])
-      max_y[i, k] <- max(y[i, 1:T, k]);
+      max_y[i, k] = max(y[i, 1:T, k]);
+  }
+  for (k in 1:7) {
+    num_obs_site[k] = 0;
+    for (i in 1:R)
+      num_obs_site[k] = num_obs_site[k] + (y[i, 1, k] != -1);
   }
 }
 
 parameters {
-  vector[7] alpha_lam;
+  vector<upper=7>[7] alpha_lam; // Constraint for stability
   vector[7] beta;
-  vector[R] eps_raw;
+  vector<upper=7>[R] eps_raw;   // Constraint for stability
```

Comment:
I think the instability is mainly due to the insufficient information of the data used in the book (this is discussed in the book). With this data, lambda can be very huge when both of p[1] and p[2] become very small during the calculation of bivariate Poisson.